cdap-app-fabric/src/test/java/co/cask/cdap/metadata/FieldLineageAdminTest.java

```diff
@@ -0,0 +1,242 @@
+
+    FieldOperationInput input = FieldOperationInput.of(Collections.singletonList(InputField.of("read", "offset")));
+    FieldOperationOutput output = FieldOperationOutput.of(Collections.singletonList("offset"));
+    fieldOperationInfos.add(new FieldOperationInfo("normalize", "normalizing offset", input, output));
+
+    inputFields = new ArrayList<>();
+    inputFields.add(InputField.of("normalize", "offset"));
+    inputFields.add(InputField.of("parse", "name"));
+    inputFields.add(InputField.of("parse", "address"));
+    inputFields.add(InputField.of("parse", "zip"));
+
+    input = FieldOperationInput.of(inputFields);
+    output = FieldOperationOutput.of(endPoint2);
+    fieldOperationInfos.add(new FieldOperationInfo("write", "writing file", input, output));
+
+    expectedInfos.add(new ProgramFieldOperationInfo(programInfos, fieldOperationInfos));
+    Assert.assertNotNull(incomings);
+    // converting to set because ordering in different versions of operations is not guaranteed
+    Assert.assertEquals(expectedInfos, new HashSet<>(incomings));
+  }
+
+  @Test
+  public void testTopologicalSortingOfOperations() {
+    Set<Operation> operations = new HashSet<>();
+
+    // read---->parse---->normalize--->write
+    ReadOperation read = new ReadOperation("read", "read descr", EndPoint.of("ns", "input"), "offset", "body");
+    operations.add(read);
+
+    TransformOperation parse = new TransformOperation("parse", "parse descr",
+                                                      Collections.singletonList(InputField.of("read", "body")), "name",
+                                                      "address");
+    operations.add(parse);
+
+    TransformOperation normalize = new TransformOperation("normalize", "normalize descr",
+                                                          Collections.singletonList(InputField.of("parse", "address")),
+                                                          "address");
+    operations.add(normalize);
+
+    List<InputField> writeInputs = new ArrayList<>();
+    writeInputs.add(InputField.of("parse", "name"));
+    writeInputs.add(InputField.of("normalize", "address"));
+    WriteOperation write = new WriteOperation("write", "write descr", EndPoint.of("ns", "output"), writeInputs);
+    operations.add(write);
+
+    FieldLineageAdmin admin = new FieldLineageAdmin(new NoopFieldLineageReader());
+    List<Operation> topologicallySortedOperations = admin.getTopologicallySortedOperations(operations);
+    assertBefore(topologicallySortedOperations, read, parse);
+    assertBefore(topologicallySortedOperations, parse, normalize);
+    assertBefore(topologicallySortedOperations, normalize, write);
+    assertBefore(topologicallySortedOperations, read, write);
+
+    // read----------------------write
+    //   \                      /
+    //    ----parse---normalize
+
+    operations = new HashSet<>();
```

Comment:
instead of using a HashSet, I would use a LinkedHashSet and also test multiple orderings. For example, when parse is first, when normalize is first, etc.

Reply:
I think it's better to break these up into separate methods. testLinearTopologicalSort, testBranchingTopologicalSort, etc. If one breaks, it will be easier to see what is wrong.

Reply:
Is it possible to have disjoint branches passed in? For example:
```
A -> B

C -> D
```

If so, please add a test for it.

Reply:
Added tests for multiple orderings and broken down the topological sorting in multiple test cases. Added case for disjoint branches too.